lib/appcues_flutter.dart

```diff
@@ -203,4 +220,78 @@ class Appcues {
     return await _methodChannel
         .invokeMethod('didHandleURL', {'url': url.toString()});
   }
+
+  // runs every time the SemanticsNode tree updates, capturing the known
+  // layout information that can be used for Appcues element targeting
+  static void _semanticsChanged() {
+    // clear out any previously captured elements in the view
+    _methodChannel.invokeMethod('resetTargetElements');
+
+    var rootSemanticNode = RendererBinding
+        .instance.pipelineOwner.semanticsOwner?.rootSemanticsNode;
+
+    if (rootSemanticNode != null) {
+
+      // this function runs on each node in the tree, looking for
+      // identifiable elements.
+      bool visitor(SemanticsNode node) {
+
+        // by default, we use the generated label, if non-empty
+        var identifier = node.label;
+        var tags = node.tags;
+
+        // look through tags for a more specific AppcuesView identifier and
+        // use that if possible.
+        if (tags != null) {
+          for (var tag in tags) {
+            if (tag is AppcuesView) {
+              identifier = tag.identifier;
+            }
+          }
+        }
+
+        if (identifier.isNotEmpty) {
+
+          // the SemanticsNode rect is in local coordinates. This helper
+          // will recursively walk the ancestors and transform the rect
+          // into global coordinates for the screen.
+          Rect transformToRoot(Rect rect, SemanticsNode? node) {
+            var transform = node?.transform;
+            if (transform == null) {
+              return rect;
+            }
+
+            var transformed = rect;
+            var offset = MatrixUtils.getAsTranslation(transform);
+            if (offset != null) {
+              transformed = rect.shift(offset);
+            }
+
+            return transformToRoot(transformed, node?.parent);
+          }
+
+          // do the transform to global coordinates
+          var rect = transformToRoot(node.rect, node);
+
+          // pass this target element through to the native side to capture
+          // in the current known set of views for targeting
+          _methodChannel.invokeMethod('targetElement', {
+            'x': rect.left,
+            'y': rect.top,
+            'width': rect.width,
+            'height': rect.height,
+            'type': 'SemanticsNode',
+            'identifier': identifier,
+          });
```

Comment:
and this is the key spot where we push view info through to the native side as we find identifiable items

Reply:
My (unverified and non-expert) instinct is that there might be a decent amount of overhead calling `invokeMethod`, so if that's true, batching all the elements into a list that we push once might be worthwhile?

Reply:
batched into array here https://github.com/appcues/appcues-flutter-plugin/pull/43/commits/cc8768cbf7e700114e94f5e1815350465a6c72da - did simplify things a bit, which is nice